lib/submit-request.js

```diff
@@ -204,6 +204,13 @@ SubmitRequest.prototype.commit = function(callback) {
       };
     }
 
+    if (request.op.op && request.op.op.length === 0 && request._fixupOps.length === 0) {
+      // The op is a no-op, either because it was submitted as such, or - more
+      // likely - because it was transformed into one. Let's avoid committing it
+      // and tell the client.
+      return callback(request.noOpError());
```

Comment:
This check has to happen at some point after the [op is transformed](https://github.com/share/sharedb/blob/945dc50ebc93d53e7306c522226bd804a802bc49/lib/submit-request.js#L152), and before the [op is committed](https://github.com/share/sharedb/blob/945dc50ebc93d53e7306c522226bd804a802bc49/lib/submit-request.js#L208).

I don't think I have a strong opinion on when we return this, so long as it's at either one of these extremes(?).

The benefit of doing it just before `commit()` is that it's a bit more directly linked to the fact that commit is about to be skipped. It does mean that all the middleware will be fired, though (which may or may not be a good/expected thing?).

Reply:
After some discussion, we decided to move just after transform, before the "apply" hook.

The "commit" hook right now means that sharedb core is going to call the DB adapter's `commit` method, and leaving the no-op early return here would have meant a violation of that behavior. By moving it earlier, that behavior is preserved.